### PR TITLE
Add shelter location filter (DEV-1078)

### PR DESCRIPTION
**Pull request by mikefeldberg**
DEV-1078

## Summary by Sourcery

Implement a new feature to filter shelters by location using geographic coordinates. Enhance the Shelter model with a geometry field to support this functionality and update the save method accordingly. Add tests to verify the correct behavior of the location filter.

New Features:
- Introduce a shelter location filter to allow querying shelters based on geographic coordinates.

Enhancements:
- Add a geometry field to the Shelter model to store geographic coordinates and update the save method to populate this field.

Tests:
- Add tests for the new shelter location filter to ensure it correctly filters shelters based on proximity to a given location.

### CODE REVIEW

File: apps/betterangels-backend/shelters/models.py [hunk 1]
```diff
@@ -171,6 +173,7 @@ class Shelter(BaseModel):
     name = models.CharField(max_length=255)
     organization = models.ForeignKey(Organization, on_delete=models.SET_NULL, blank=True, null=True)
     location = PlacesField(blank=True, null=True)
+    geometry = GeometryField(srid=4326, blank=True, null=True)
```
Comment (vecchp) [inline apps/betterangels-backend/shelters/models.py:176]: https://docs.djangoproject.com/en/5.1/ref/contrib/gis/model-api/#geography 

you'll want to set the field to `geography=true`

File: apps/betterangels-backend/shelters/types.py [hunk 2]
```diff
@@ -155,7 +157,25 @@ class FunderType:
...
+@strawberry_django.filters.filter(Shelter)
+class ShelterLocationFilter:
+    @strawberry_django.filter_field
+    def geometry(self, queryset: QuerySet, value: Optional[GeometryInput], prefix: str) -> Tuple[QuerySet[Shelter], Q]:
+        if value is None:
+            return queryset, Q()
+
+        user_location = Point(x=value.longitude, y=value.latitude, srid=4326)
+
+        return queryset.annotate(distance=Distance("geometry", user_location)).order_by("distance"), Q()
```
[the remark this answers is not in the document]
Comment (mikefeldberg) [reply apps/betterangels-backend/shelters/types.py:175]: do we want to hardcode max distance like that? Maybe we expand the filter to also take a distance parameter?

File: apps/betterangels-backend/shelters/models.py [hunk 3]
```diff
@@ -248,6 +251,17 @@ class Meta:
     def __str__(self) -> str:
         return self.name
 
+    def save(self, *args: Any, **kwargs: Any) -> None:
+        latitude = self.location.latitude if self.location else None
+        longitude = self.location.longitude if self.location else None
+
+        if latitude is not None and longitude is not None:
+            self.geometry = Point(float(longitude), float(latitude), srid=4326)
```
[the remark this answers is not in the document]
Comment (mikefeldberg) [reply apps/betterangels-backend/shelters/models.py:259]: which part exactly may not be necessary?

File: apps/betterangels-backend/shelters/models.py (hunk 1 above)
Comment (mikefeldberg) [reply apps/betterangels-backend/shelters/models.py:176]: good point

**Review by vecchp — APPROVED** [review]

**Pull request merged**